connector/spanmetricsconnectorv2/internal/aggregator/histogram/explicit.go

```diff
@@ -0,0 +1,179 @@
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package histogram // import "github.com/elastic/opentelemetry-collector-components/connector/spanmetricsconnectorv2/internal/aggregator/histogram"
+
+import (
+	"sort"
+	"time"
+
+	"github.com/elastic/opentelemetry-collector-components/connector/spanmetricsconnectorv2/config"
+	"github.com/elastic/opentelemetry-collector-components/connector/spanmetricsconnectorv2/internal/model"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/pdatautil"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+)
+
+// metricUnitToDivider gives a value that could used to divide the
+// nano precision duration to the required unit specified in config.
+var metricUnitToDivider = map[config.MetricUnit]float64{
+	config.MetricUnitNs: float64(time.Nanosecond.Nanoseconds()),
+	config.MetricUnitUs: float64(time.Microsecond.Nanoseconds()),
+	config.MetricUnitMs: float64(time.Millisecond.Nanoseconds()),
+	config.MetricUnitS:  float64(time.Second.Nanoseconds()),
+}
+
+// ExplicitBounds is a representation of explict bound histogram for
+// calculating histograms for span durations.
+type ExplicitBounds struct {
+	// TODO (lahsivjar): Attribute hash collisions are not considered
+	datapoints map[metricKey]map[[16]byte]*explicitHistogramDP
+	timestamp  time.Time
+}
+
+// NewExplicitBounds creates a new instance of explicit bounds histogram.
+func NewExplicitBounds() *ExplicitBounds {
+	return &ExplicitBounds{
+		datapoints: make(map[metricKey]map[[16]byte]*explicitHistogramDP),
+		timestamp:  time.Now(),
+	}
+}
+
+// Add adds a datapoint for a given metric definition to the histogram.
+func (h *ExplicitBounds) Add(
+	md model.MetricDef,
+	srcAttrs pcommon.Map,
+	spanDuration time.Duration,
+) error {
+	filteredAttrs := pcommon.NewMap()
+	for _, definedAttr := range md.Attributes {
+		if srcAttr, ok := srcAttrs.Get(definedAttr.Key); ok {
+			srcAttr.CopyTo(filteredAttrs.PutEmpty(definedAttr.Key))
+			continue
+		}
+		if definedAttr.DefaultValue.Type() != pcommon.ValueTypeEmpty {
+			definedAttr.DefaultValue.CopyTo(filteredAttrs.PutEmpty(definedAttr.Key))
+		}
+	}
+
+	// If all the configured attributes are not present in source
+	// metric then don't count them.
+	if filteredAttrs.Len() != len(md.Attributes) {
+		return nil
+	}
```

Comment:
[**For reviewers**] I plan to introduce another abstraction when the summaries and exp histograms support will be added. This bit of code will move to that abstraction layer which will handle all types (explicit bounds histogram, expo histograms, summaries) and all the explicit types (like `ExplicitBounds`) would be unexported.